Split build workflows for different platforms

joyeecheung opened the pull request:
Using one workflow to build for multiple platforms can lead to timeout errors if one of them takes too long (especially when there's emulation), and no images would be published even when only one of them fails.

This split the build workflows into different files based on the platform that publish the images to platform-specific nightly channels, and add a new workflow to merge the platform-specific nightlies into one nightly channel, so that
when one platform fails to build, others can still be updated. This also allows us to build the arm64 images without emulation which would be less likely to time out.

joyeecheung committed:
Split build workflows for different platforms

Using one workflow to build for multiple platforms can lead
to timeout errors if one of them takes too long (especially when
there's emulation), and no images would be published even when
only one of them fails.

This split the build workflows into different files based on the
platform that publish the images to platform-specific nightly
channels, and add a new workflow to merge the platform-specific
nightlies into one nightly channel, so that
when one platform fails to build, others can still be updated.
This also allows us to build the arm64 images without emulation
which would be less likely to time out.

bnb commented on .github/workflows/docker-build-and-publish-linux-amd64.yml:
```diff
@@ -0,0 +1,27 @@
+jobs:
+  docker:
+    if: ${{ github.ref == 'refs/heads/main' }}
+    runs-on: ubuntu-latest
```
since we're pinning to `ubuntu-24.04-arm64` for arm64, maybe we pin to `ubuntu-24.04` for consistency?

```suggestion
    runs-on: ubuntu-24.04
```

joyeecheung replied:
Was only using `ubuntu-24.04-arm64` because there's no `ubuntu-latest-arm64` (not sure why 🤔) but I think we can be more consistent on the other side..

bnb commented on .github/workflows/docker-merge.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      # Merge the multiple architecture-specific images into a single multi-arch image.
+      # If images for other architectures do not exist in the registry yet, this step will fail.
+      # It will only succeed when at least one image for each architecture has been built and pushed.
+      # If images for other architectures are being built and the ones in the registry
+      # are old, it only updates the nightly tag for the architecture that triggered this
+      # workflow. When images for other architectures are built successfully later, they will
+      # trigger this workflow again and merge the updated images.
+      - name: Create multi-arch manifest
+        run: |
+          docker buildx imagetools create \
+            -t ${REGISTRY}/${IMAGE}:nightly \
+            ${REGISTRY}/${IMAGE}:nightly-linux-amd64 \
+            ${REGISTRY}/${IMAGE}:nightly-linux-arm64
+
+      - name: Verify multi-arch manifest
+        run: |
+          docker buildx imagetools inspect ${REGISTRY}/${IMAGE}:nightly
```
Post-merge, should we be publishing this multi-arch image?

joyeecheung replied:
I think the `docker buildx imagetools create` step above should do that https://docs.docker.com/reference/cli/docker/buildx/imagetools/create/ (it's the default unless `--dry-run` is used)

bnb replied:
If it's getting published with the `docker buildx imagetools create` step, does it still make sense to `docker buildx imagetools inspect` once that's completed? 🤔

joyeecheung replied:
It's mostly for tracing image updates in the workflow logs, e.g. https://github.com/joyeecheung/devcontainer/actions/runs/18767496816/job/53545272640 + https://github.com/joyeecheung/devcontainer/actions/runs/18767479326/job/53545229049 shows what got updated with which SHA

joyeecheung committed:
make the workflow runnable from forks

joyeecheung committed:
fix typo in arm64 workflow

joyeecheung commented:
Updated the workflow files a bit so that the dockerhub image name and the nightly branch are configurable via repository variables, so that I could test it in my fork. I've added the corresponding variables to this repo too.

<img width="793" height="190" alt="Screenshot 2025-10-23 at 17 38 10" src="https://github.com/user-attachments/assets/6c1923d4-8c9e-4752-8d23-f7ff0f6fcd56" />

The test builds from my fork are in here https://hub.docker.com/r/joyeecheung/node-devcontainer/tags

joyeecheung commented:
@bnb Can you take a look again? Thanks!

bnb commented:
Looks good to me other than https://github.com/nodejs/devcontainer/pull/14#discussion_r2460930121 but will approve on the case that there's no change needed ✅
